R/http.R

```diff
@@ -1,7 +1,170 @@
+# Environment in which cookies will be stored. Cookies are expected to survive
+# the duration of the R session, but are not persisted outside of the R
+# session.
+.cookieStore <- new.env(parent=emptyenv())
+
 userAgent <- function() {
   paste("rsconnect", packageVersion("rsconnect"), sep="/")
 }
 
+getCookieHost <- function(requestURL){
+  host <- requestURL$host
+  port <- requestURL$port
+  if (nchar(port) > 0){
+    port <- sub("^:", "", port)
+    # By my reading of the RFC, we technically only need to include the port #
+    # in the index if the host is an IP address. But here we're including the
+    # port number as a part of the host whether using a domain name or IP.
+    # Erring on the side of not sending the cookies to the wrong services
+    host <- paste(host, port, sep=":")
+  }
+}
+
+# Parse out the raw headers provided and insert them into the cookieStore
+# NOTE: Domain attribute is currently ignored
+# @param requestURL the parsed URL as returned from `parseHttpUrl`
+# @param cookieHeaders a vector of characters strings representing the raw
+#   Set-Cookie header value with the "Set-Cookie: " prefix omitted
+storeCookies <- function(requestURL, cookieHeaders){
+  cookies <- lapply(cookieHeaders, function(co){ parseCookieHeader(requestURL, co) })
+
+  # Flatten out since some headers might contain multiple cookies
+  cookies <- unlist(cookies, recursive=FALSE)
+
+  # Filter out invalid cookies (which would return as NULL)
+  cookies <- Filter(Negate(is.null), cookies)
+
+  host <- getCookieHost(requestURL)
+
+  hostCookies <- NULL
+  if (!exists(host, .cookieStore)){
+    # Create a new data frame for this host
+    hostCookies <- data.frame(
+      path=character(0),
+      name=character(0),
+      value=character(0),
+      secure=logical(0),
+      expires=character(0),
+      stringsAsFactors = FALSE
+    )
+  } else {
+    hostCookies <- get(host, envir=.cookieStore)
+  }
+
+  lapply(cookies, function(co){
+    # Remove any duplicates
+    # RFC says duplicate cookies are ones that have the same domain, name, and path
+    hostCookies <<- hostCookies[!(co$name == hostCookies$name & co$path == hostCookies$path),]
+
+    # append this new cookie on
+    hostCookies <<- rbind(co, hostCookies, stringsAsFactors=FALSE)
+  })
+
+  # Save this host's cookies into the cookies store.
+  assign(host, hostCookies, envir=.cookieStore)
+}
+
+# Parse out an individual cookie
+# @param requestURL the parsed URL as returned from `parseHttpUrl`
+# @param cookieHeader the raw text contents of the Set-Cookie header with the
+#   header name omitted. May contain multiple comma-separated cookies
+parseCookieHeader <- function(requestURL, cookieHeader){
+  cookieStrs <- trimws(strsplit(cookieHeader, ",", fixed=TRUE)[[1]])
```

Comment:
I think `trimws` was only added in R 3.3 or 3.4 so you might want to just use e.g.

    gsub("^\\s*|\\s*$", "", text)